### ISSUE TITLE
Slow IO performance with VirtualBox on host with ext4

### ISSUE THREAD

**Issue by outofrange** [oracle/vagrant-projects#280]
We use the `OracleDatabase` (18.4.0-XE) vagrant template to set up our development database with OracleXE (18c-1.0-1).
I noticed that the performance of the created VirtualBox VM is far worse on my machine compared to those of my colleagues; while their `vagrant up` (mainly updating packages, installing Oracle and importing a dump) takes something between 30 minutes to less than two hours on Windows, it takes more than 14 hours on my Linux system.

After some investigation, I narrowed it down to very slow IO performance, seeing with `iotop` that `jbd2`, ext4's journaling process, constantly uses somewhere between 95 and >99% of my IO.
First thing I found was [a blog warning about some Kernel regression](https://www.percona.com/blog/2019/11/12/watch-out-for-disk-i-o-performance-issues-when-running-ext4/), but this didn't quite matched my situation, since it should've been fixed already; nevertheless, it got me thinking about other possible reasons and solutions, so I don't have to (try to) disable journaling on the paritition in use.

Eventually, I found another [blog](https://www.electricmonk.nl/log/2016/03/14/terrible-virtualbox-disk-performance/) talking about IO performance problems when the VM's SATA controller isn't using host's IO cache (which it doesn't by default).

Indeed, when configured in `Vagrantfile` with `v.customize ["storagectl", :id, "--name", "SATA Controller", "--hostiocache", "on"]`, VM provisioning only takes 40 minutes.

While this seems to work fine, there is [a reason](https://forums.virtualbox.org/viewtopic.php?p=206796#p206796) why it isn't enabled by default, so I'm not sure if that's the best solution or if I should submit a PR for enabling hostiocache.

Other approaches I considered, but haven't tried:

- not using ext4 (not sure if behaviour is different with i.e. xfs)
- not using journaling (I'd rather use journaling, for obvious reasons)
- changing Oracle parameters `filesystemio_options` and `disk_asynch_io` (updating packages with `yum` felt slow too, so it looks like it's not a problem rooted in Oracle and its configuration)

I was able to reproduce the problem on two different machines - one of them being a laptop with an encrypted nvme ssd, the other one a regular PC with a non-encrypted HDD.
The following is true for both of them:

- Arch based distros running Linux kernel 5.7.7-arch1-1
- ext4 mounted with options `defaults,relatime`
- Vagrant 2.2.9
- VirtualBox 6.1.10r138449

I'm curious if someone else has the same issue and if enabling `hostiocache` is the way to go. If so, I can provide a small PR changing it.

**Comment by AmedeeBulle** [comment]
Thank you for the comprehensive report.

To avoid misunderstanding, can you confirm your are talking about ext4 on the host machine and not in the guest?  
(Oracle boxes are using xfs, but I want to double check...)

**Comment by outofrange** [comment]
Yes - ext4 on host; guest is using xfs.

**Comment by AmedeeBulle** [comment]
Oops -- Apologies, deleted my comment in between, as I just saw _Host_ in subject...

**Comment by gvenzl** [comment]
Hey @scoter-oracle, are there any explanations why this takes 14 hours on that machine? I just tested it on my Oracle Linux box and the provisioning of the VM takes about 21 minutes only.

It's probably a good exercise to see whether we should set some additional parameters to speed up these VMs.

@outofrange, 40 minutes still seems rather long. Is that for just provisioning the VM, i.e. `vagrant up` or are there additional steps that are included in these 40 minutes? I see you are referring to "and importing a dump" above.

**Comment by Djelibeybi** [comment]
@gvenzl Oracle Linux defaults to XFS not ext4. We haven't used ext4 as the default file system for _years_. So to answer @outofrange's question, yes this is very different on XFS and we would strongly recommend it over ext4 these days.

**Comment by outofrange** [comment]
@gvenzl oh yes, I'm not surprised about the 40 minutes from first `vagrant up` to Oracle being up and ready, that's fine and what I'm expecting - we're using `impdb` to import a subset of tables, users, ... from a dump taken from our on-prem test system; the dump itself is 13GB - I'm not sure how much smaller the actually imported subset is. A wild guess would be somewhere between 7GB and 10GB, leaving us with enough room to stay under XE's size limitations during development.
Without having taken any exact measurements, `DATABASE_EXPORT/SCHEMA/TABLE/INDEX/INDEX` is the step during the import that takes by far the longest.

I would be happy to do some benchmarks and provide a small, self contained example for this issue. Do you have a recommendation for what exactly I should execute/measure? Would this be helpful?


And to be clear, I don't think that the root cause is with OracleXE, Oracle Linux or your Vagrantfile in particular.
But if there's a reasonable change in the scope of this repository, to alleviate the issue of, say, XFS on LVM on the guest side, on top of ext4 on the host (if _that's_ the cause), I'm all for that :)
Otherwise, I'm fine with @Djelibeybi recommendation of XFS over ext4 and having it documented here so others can find it. Maybe in README too?

**Comment by outofrange** [comment]
Btw, enabling `hostiocache` doesn't increase the performance as dramatically on my laptop as it does on my PC.
But I forgot to mention that there is a significant difference between my two machines: my laptops nvme SSD is encrypted (so I'm not that surprised that it's still stuck on processing some index), while my HDD is not.

Still, performance doesn't feel "right" on my laptop either for having a SSD, but I can't see `jbd2` hogging IO, so maybe it's a different issue. I'll try to dig deeper on that too, but I'll focus on my setup without additional encryption for now.

**Comment by AmedeeBulle** [comment]
I compared the time to run a `vagrant up` for the 19c project in the following environments:

**Providers**:
- VirtualBox 6.0.22r137980
- VirtualBox 6.1.10r138449

**Oracle Linux 7 with the following kernels**:
- UEK5 -- 4.14.35-1902.303.5.3
- UEK6 -- 5.4.17-2011.3.2.1

**File systems**:
- Ext4 on LVM
- XFS on LVM

**Platforms**:
- Bare Metal (workstation)
- Oracle Linux Virtualization Manager VM (KVM with qcow2 storage)
- Oracle Cloud Infrastructure

On each platform the results are pretty consistent; the difference between the slowest and the fastest run is less than 5% with no particular pattern (that is: the variation is most probably due to system activity during the run).

I am not sure what is going pear-shaped in your environment... (your bleeding edge kernel?)

While it doesn't answer the question, since your host in Linux anyway you might try the `libvirt` vagrant provider to get the benefits of KVM virtualisation.

**Comment by outofrange** [comment]
@AmedeeBulle Thank you for checking!

I hope I'll find time to do some further testing on the weekend. I want to try an older kernel, libvirt and xfs. And maybe using `noatime` instead of `relatime` for ext4.

Also, I did some `dd` write tests in my VM - while this happened quite fast (even `sync`ing) both with and without `hostiocache`, I noticed `jbd2` spinning up only when `hostiocache` is disabled, and using ~98% of my IO after `sync` finished.
Unfortunetely, I'm not sure yet how to measure write times including writeback of the cache. Maybe `fio` will help me get actual benchmarks.
